### DISCUSSION TITLE
Building non provided components

### DISCUSSION THREAD

**Discussion by imjakechapman** [segmentio/evergreen#1141] [Q&A]
Just found Evergreen and first I have to say Segment team nailed the visual design of the library.

I'm trying to find examples of how to properly build out components that aren't provided by the library itself. I understand using `<Pane>` and `<Card>` is the way to go, but what is the best approach to take for styling those? css-in-js? css-modules? How do we maintain the consistency of things like theme palette when creating our own components? Maybe I missed something that the library provides `:root` css vars to use within our css. Also what are your approaches to accessibility for keyboard navigation when building components that aren't provided in the library at Segment?

Looking forward to trying this out once I get past this first hurdle.

**Comment by akleiner2** [comment]
Hey @imjakechapman, appreciate the kind words - glad you're excited about using Evergreen for an upcoming project! To answer your question(s), it might be worth talking a bit about some of the underlying architecture first. 

- Evergreen components (including `<Pane />` and `<Card />`) are all built on [ui-box](https://github.com/segmentio/ui-box). ui-box is a run-time, atomic CSS solution that targets base-styles for components (i.e. anything that doesn't target a pseudo-state / pseudo-selector of a component).
- For targeting those pseudo-states, you'll need to use some other css-in-js solution. Evergreen uses [glamor](https://github.com/threepointone/glamor) to generate a hashed classname. You could, in theory, also use Emotion, CSS Modules, Sass Modules, or some other standalone solution that outputs a classname (either at runtime or build-time). `<Pane />` and `<Card />` accept arbitrary classnames which allow that to happen. I'm unsure how Evergreen would play with `styled-components`, but might be worth a try. 

For keyboard accessibility - it very much depends on what you're looking to build. Evergreen has some of the recommendations from WCAG 2.1 implemented for things like Dialog, Popover, Button, etc. But, if you're looking to build something that's not in Evergreen, we unfortunately don't provide composable keyboard primitives to help ease the process (although, would be happy to accept PRs that introduce it into the lib!)

Hope that answers your question(s)!
